vast/losses/losses.py

```diff
@@ -137,30 +137,27 @@ def __call__(self, logit_values, target, sample_weights=None):
 
 
 class objectoSphere_loss:
-    def __init__(self, batch_size, knownsMinimumMag=50.0):
+    def __init__(self, batch_size = None, knownsMinimumMag=50.0): # batchsize ignored for backward compatibility
         self.knownsMinimumMag = knownsMinimumMag
-        self.knownsMinimumMag_tensor = (
-            tools.device(torch.ones(batch_size)) * self.knownsMinimumMag
-        )
-        self.batch_size = batch_size
 
     def __call__(self, features, target, sample_weights=None):
+        # compute feature magnitude
         mag = features.norm(p=2, dim=1)
-        if features.shape[0] != self.batch_size:
-            self.knownsMinimumMag_tensor = (
-                tools.device(torch.ones(features.shape[0])) * self.knownsMinimumMag
-            )
-            self.batch_size = features.shape[0]
-        # For knowns magnitude minus \beta is loss
-        mag_diff_from_ring = torch.clamp(self.knownsMinimumMag_tensor - mag, min=0.0)
-        loss = tools.device(torch.zeros(self.batch_size))
+        # For knowns we want a certain magnitude
+        mag_diff_from_ring = torch.clamp(self.knownsMinimumMag - mag, min=0.0)
+
+        # Loss per sample
+        loss = tools.device(torch.zeros(features.shape[0]))
         known_indexes = target != -1
         unknown_indexes = ~known_indexes
+        # knowns: punish if magnitude is inside of ring
         loss[known_indexes] = mag_diff_from_ring[known_indexes]
+        # unknowns: punish any magnitude
         loss[unknown_indexes] = mag[unknown_indexes]
         loss = torch.pow(loss, 2)
         if sample_weights is not None:
             loss = sample_weights * loss
+        # todo: return average loss
```

Comment:
The reason all loss functions inside vast are designed to return a per sample loss is so that if someone needs to do hard negative mining based on per sample loss they can do it outside without modifying the code here. Though, I do see the need for a reduction function like in most loss functions in torch, which can make this much easier. Will look into this.

Reply:
I have made the required changes, now for any function, you can pass a `reduction` argument with choices `none`, `sum` or `mean`.

Reply:
Great, thanks. Do you think, this one can be merged now?

Reply:
I am merging it for now, but should likely get back to it at some point. The way I implemented `reduction` it is expected as a parameter in the forward pass rather than being set in the class constructor. The easiest way would be through a generalized loss class and using it as a parent for all classes. I am open to suggestions but merging this for now since it doesn't break anything.